### PR TITLE
Configuration tests

### PR DESCRIPTION
**Pull request by t-mon**

### CODE REVIEW

File: libguh-core/guhconfiguration.cpp [hunk 1]
```diff
@@ -164,9 +164,14 @@ void GuhConfiguration::setServerName(const QString &serverName)
 
     GuhSettings settings(GuhSettings::SettingsRoleGlobal);
     settings.beginGroup("guhd");
-    settings.setValue("name", serverName);
-    settings.endGroup();
-    emit serverNameChanged();
+    if (settings.value("name").toString() == serverName) {
+        qCDebug(dcApplication()) << "Configuration: Server name unchainged.";
```
Comment (mzanetti) [inline libguh-core/guhconfiguration.cpp]: unchanged

File: libguh-core/guhconfiguration.cpp [hunk 2]
```diff
@@ -182,9 +187,14 @@ void GuhConfiguration::setTimeZone(const QByteArray &timeZone)
 
     GuhSettings settings(GuhSettings::SettingsRoleGlobal);
     settings.beginGroup("guhd");
-    settings.setValue("timeZone", timeZone);
-    settings.endGroup();
-    emit timeZoneChanged();
+    if (settings.value("timeZone").toByteArray() == timeZone) {
+        qCDebug(dcApplication()) << "Configuration: Time zone unchainged.";
```
Comment (mzanetti) [inline libguh-core/guhconfiguration.cpp]: unchanged

File: libguh-core/guhconfiguration.cpp [hunk 3]
```diff
@@ -200,9 +210,14 @@ void GuhConfiguration::setLocale(const QLocale &locale)
 
     GuhSettings settings(GuhSettings::SettingsRoleGlobal);
     settings.beginGroup("guhd");
-    settings.setValue("language", locale.name());
-    settings.endGroup();
-    emit localeChanged();
+    if (settings.value("language").toString() == locale.name()) {
+        qCDebug(dcApplication()) << "Configuration: Language unchainged.";
```
Comment (mzanetti) [inline libguh-core/guhconfiguration.cpp]: unchanged

File: tests/auto/configurations/testconfigurations.cpp [hunk 4]
```diff
@@ -62,19 +68,52 @@ void TestConfigurations::getConfigurations()
...
     QVERIFY(timeZones.count() > 0);
     QVERIFY(timeZones.contains("America/Toronto"));
     QVERIFY(timeZones.contains("Europe/Vienna"));
+    QVERIFY(timeZones.contains("Africa/Dakar"));
 
-    // Set current timezone (Europe/Vienna)
+    // Get current configurations
+    QVariantMap basicConfigurationMap = loadBasicConfiguration();
+
+    // Set timezone unchainged
```
Comment (mzanetti) [inline tests/auto/configurations/testconfigurations.cpp]: again

File: tests/auto/configurations/testconfigurations.cpp [hunk 5]
```diff
@@ -62,19 +68,52 @@ void TestConfigurations::getConfigurations()
...
     params.clear(); response.clear(); configurations.clear();
-    params.insert("timeZone", "Europe/Vienna");
+    params.insert("timeZone", basicConfigurationMap.value("timeZone").toString());
     response = injectAndWait("Configuration.SetTimeZone", params);
     verifyConfigurationError(response);
 
+    // Check notification not emitted
+    notificationSpy.wait(200);
```
Comment (mzanetti) [inline tests/auto/configurations/testconfigurations.cpp]: perhaps better to use wait() without argument (the default 5 secs). If jenkins is under high load 200ms could make it randomly fail

File: tests/auto/configurations/testconfigurations.cpp [hunk 6]
```diff
@@ -62,19 +68,52 @@ void TestConfigurations::getConfigurations()
...
+    QVERIFY2(configurationChangedNotifications.count() == 0, "Got Configuration.BasicConfigurationChanged notification but should have not.");
+
+    // Set new timezone (Africa/Dakar)
+    QString newTimeZone("Africa/Dakar");
+    params.clear(); response.clear(); configurations.clear(); notificationSpy.clear();
+    params.insert("timeZone", newTimeZone);
+    response = injectAndWait("Configuration.SetTimeZone", params);
+    verifyConfigurationError(response);
+
+    notificationSpy.wait(200);
```
Comment (mzanetti) [inline tests/auto/configurations/testconfigurations.cpp]: -200

File: tests/auto/configurations/testconfigurations.cpp [hunk 7]
```diff
@@ -109,33 +148,143 @@ void TestConfigurations::testTimeZones()
...
+    QVariantMap params; QVariant response; QVariantList configurationChangedNotifications;
+    params.insert("serverName", serverName);
     response = injectAndWait("Configuration.SetServerName", params);
     verifyConfigurationError(response);
 
-    configurations = injectAndWait("Configuration.GetConfigurations").toMap().value("params").toMap();
-    QString newServerName = configurations.value("basicConfiguration").toMap().value("serverName").toString();
-    QVERIFY(newServerName == "Test server");
+    // Check notification not emitted
+    notificationSpy.wait(200);
```
Comment (mzanetti) [inline tests/auto/configurations/testconfigurations.cpp]: -200?

File: tests/auto/configurations/testconfigurations.cpp [hunk 8]
```diff
@@ -109,33 +148,143 @@ void TestConfigurations::testTimeZones()
...
+        notificationSpy.wait(200);
+        QVariantList languageChangedNotifications = checkNotifications(notificationSpy, "Configuration.LanguageChanged");
+
+        // If the language did not change no notification should be emited
+        if (basicConfigurationMap.value("language").toString() == languageVariant.toString()) {
+            QVERIFY2(languageChangedNotifications.count() == 0, "Got Configuration.LanguageChanged notification but should have not.");
+        } else {
+            QVERIFY2(languageChangedNotifications.count() == 1, "Should get only one Configuration.LanguageChanged notification");
+
+            // TODO: verify notification
```
Comment (mzanetti) [inline tests/auto/configurations/testconfigurations.cpp]: now or never, I guess :D

**Comment by t-mon** [comment]
Jenkins, retest this please.

**Comment by mzanetti** [comment]
there are still some wait(200). dunno... you decide. the rest looks good to me.

**Pull request merged**